### PR TITLE
[modbus] handle floating point NaN and inf gracefully

### LINKED ISSUE #5476: [modbus] 2.4 binding crashes during swap float32
modbus stops reading registerdata

modbus 2.4

Here the error trace:
```
2019-04-17 12:25:32.779 [WARN ] [mmon.WrappedScheduledExecutorService] - Scheduled runnable ended with an exception:
java.lang.NumberFormatException: null
        at java.math.BigDecimal.<init>(BigDecimal.java:494) ~[?:?]
        at java.math.BigDecimal.<init>(BigDecimal.java:383) ~[?:?]
        at java.math.BigDecimal.<init>(BigDecimal.java:806) ~[?:?]
        at java.math.BigDecimal.valueOf(BigDecimal.java:1274) ~[?:?]
        at org.eclipse.smarthome.core.library.types.DecimalType.<init>(DecimalType.java:53) ~[?:?]
        at org.openhab.io.transport.modbus.ModbusBitUtilities.extractStateFromRegisters(ModbusBitUtilities.java:172) ~[?:?]
        at org.openhab.binding.modbus.internal.handler.ModbusDataThingHandler.onRegisters(ModbusDataThingHandler.java:664) ~[?:?]
        at org.openhab.binding.modbus.internal.handler.ModbusPollerThingHandlerImpl$ReadCallbackDelegator.lambda$0(ModbusPollerThingHandlerImpl.java:91) ~[?:?]
        at java.util.concurrent.CopyOnWriteArrayList.forEach(CopyOnWriteArrayList.java:891) ~[?:?]
        at org.openhab.binding.modbus.internal.handler.ModbusPollerThingHandlerImpl$ReadCallbackDelegator.onRegisters(ModbusPollerThingHandlerImpl.java:91) ~[?:?]
        at org.openhab.io.transport.modbus.internal.ModbusLibraryWrapper.invokeCallbackWithResponse(ModbusLibraryWrapper.java:279) ~[?:?]
        at org.openhab.io.transport.modbus.internal.ModbusManagerImpl$PollOperation.lambda$1(ModbusManagerImpl.java:163) ~[?:?]
        at org.openhab.io.transport.modbus.internal.SimpleStopWatch.timeRunnable(SimpleStopWatch.java:148) ~[?:?]
        at org.openhab.io.transport.modbus.internal.ModbusManagerImpl$PollOperation.accept(ModbusManagerImpl.java:162) ~[?:?]
        at org.openhab.io.transport.modbus.internal.ModbusManagerImpl$PollOperation.accept(ModbusManagerImpl.java:1) ~[?:?]
        at org.openhab.io.transport.modbus.internal.ModbusManagerImpl.executeOperation(ModbusManagerImpl.java:571) ~[?:?]
        at org.openhab.io.transport.modbus.internal.ModbusManagerImpl.lambda$15(ModbusManagerImpl.java:719) ~[?:?]
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) ~[?:?]
        at java.util.concurrent.FutureTask.runAndReset(FutureTask.java:308) ~[?:?]
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$301(ScheduledThreadPoolExecutor.java:180) ~[?:?]
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:294) ~[?:?]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [?:?]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [?:?]
        at java.lang.Thread.run(Thread.java:748) [?:?]
```

## Expected Behavior

If the readed register does not contain valid data, or it could not be read by any reason,
the binding should be robust against such corruped data. Right now the thread is crashing and reading of the registers is stopped.

catch the error and try it again next cycle.  After e.g. 10 invalid reads in a row set an error.

## Possible Solution
catch the error and try it again next cycle.  After e.g. 10 invalid reads in a row set an error.

## Steps to Reproduce (for Bugs)
sporadic -> maybe my wago SPS does not always like the floats

## Context
recording of temp data is interupted an I have to monitor it manualy if binding stopped working.
With the modbus 1.x binding it was running without problems 1 and a half year.

## Your Environment
openhabian 2.4

### PR DESCRIPTION
**Pull request by ssalonen**
With this fix, if the registers contain floating point NaN or infinity,
and we are parsing float32 or float_swap values, we return UNDEF instead
of crashing.

The binding updates number channels as UNDEF.

Fixes #5476

Signed-off-by: Sami Salonen <ssalonen@gmail.com>

### CODE REVIEW

**Comment by ssalonen** [comment]
This is backwards incompatible change to exported API (return value has changed).

This will break e.g. #4220, but it should be trivial enough to fix.

File: bundles/org.openhab.io.transport.modbus/src/main/java/org/openhab/io/transport/modbus/ModbusBitUtilities.java [hunk 1]
```diff
@@ -205,10 +220,10 @@ public static DecimalType extractStateFromRegisters(ModbusRegisterArray register
      * NUL byte encountered.
      *
      * @param registers list of registers, each register represent 16bit of data
-     * @param index zero based register index. Registers are handled as 16bit registers,
-     *            this parameter defines the starting register.
-     * @param length maximum length of string in 8bit characters.
-     * @param charset the character set used to construct the string.
+     * @param index     zero based register index. Registers are handled as 16bit registers,
```
Comment (davidgraeff) [inline bundles/org.openhab.io.transport.modbus/src/main/java/org/openhab/io/transport/modbus/ModbusBitUtilities.java]: Please apply the ESH formatter profile again in your eclipse installation and reformat this file

File: bundles/org.openhab.io.transport.modbus/src/main/java/org/openhab/io/transport/modbus/ModbusBitUtilities.java [hunk 2]
```diff
@@ -173,7 +183,12 @@ public static DecimalType extractStateFromRegisters(ModbusRegisterArray register
                 ByteBuffer buff = ByteBuffer.allocate(4);
                 buff.put(registers.getRegister(index + 1).getBytes());
                 buff.put(registers.getRegister(index).getBytes());
-                return new DecimalType(buff.order(ByteOrder.BIG_ENDIAN).getFloat(0));
+                try {
+                    return new DecimalType(buff.order(ByteOrder.BIG_ENDIAN).getFloat(0));
```
Comment (davidgraeff) [inline bundles/org.openhab.io.transport.modbus/src/main/java/org/openhab/io/transport/modbus/ModbusBitUtilities.java]: If there are more State types I could understand this API change. But are you sure that an `Optional` or a null value are not expressive enough?

Comment (ssalonen) [reply bundles/org.openhab.io.transport.modbus/src/main/java/org/openhab/io/transport/modbus/ModbusBitUtilities.java]: I like what you are proposing, makes it more apparent what is returned.

Will change it to use optional

File: bundles/org.openhab.io.transport.modbus/src/main/java/org/openhab/io/transport/modbus/ModbusBitUtilities.java (hunk 1 above)
Comment (ssalonen) [reply bundles/org.openhab.io.transport.modbus/src/main/java/org/openhab/io/transport/modbus/ModbusBitUtilities.java]: I have confirmed that the formatter is active but still get the same result. Where would I find up-to-date formatter definition? 

Is the only option to reset the workspace? :(

File: bundles/org.openhab.io.transport.modbus/src/main/java/org/openhab/io/transport/modbus/ModbusBitUtilities.java (hunk 2 above)
Comment (ssalonen) [reply bundles/org.openhab.io.transport.modbus/src/main/java/org/openhab/io/transport/modbus/ModbusBitUtilities.java]: Resolved in b20aab0

File: bundles/org.openhab.io.transport.modbus/src/main/java/org/openhab/io/transport/modbus/ModbusBitUtilities.java (hunk 1 above)
Comment (davidgraeff) [reply bundles/org.openhab.io.transport.modbus/src/main/java/org/openhab/io/transport/modbus/ModbusBitUtilities.java]: just chose another one and then ESH again and apply. It's just a bugged Eclipse

Comment (ssalonen) [reply bundles/org.openhab.io.transport.modbus/src/main/java/org/openhab/io/transport/modbus/ModbusBitUtilities.java]: Oh! Did not know that...

Whole modbus code base formatted in 39ff5a4.

**Pull request merged**